velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -621,182 +484,143 @@ class Divide {
     auto precision = aPrecision - aScale + bScale + scale;
     return DecimalUtil::adjustPrecisionScale(precision, scale);
   }
+
+  uint8_t aRescale_;
+  uint8_t rPrecision_;
 };
 
-std::vector<std::shared_ptr<exec::FunctionSignature>>
-decimalAddSubtractSignature() {
-  return {
-      exec::FunctionSignatureBuilder()
-          .integerVariable("a_precision")
-          .integerVariable("a_scale")
-          .integerVariable("b_precision")
-          .integerVariable("b_scale")
-          .integerVariable(
-              "r_precision",
-              "min(38, max(a_precision - a_scale, b_precision - b_scale) + max(a_scale, b_scale) + 1)")
-          .integerVariable(
-              "r_scale",
-              getResultScale(
-                  "max(a_precision - a_scale, b_precision - b_scale) + max(a_scale, b_scale) + 1",
-                  "max(a_scale, b_scale)"))
-          .returnType("DECIMAL(r_precision, r_scale)")
-          .argumentType("DECIMAL(a_precision, a_scale)")
-          .argumentType("DECIMAL(b_precision, b_scale)")
-          .build()};
+template <template <class> typename Func>
+void registerDecimalBinary(
+    const std::string& name,
+    std::vector<exec::SignatureVariable> constraints) {
+  // (long, long) -> long
+  registerFunction<
+      Func,
+      LongDecimal<P3, S3>,
+      LongDecimal<P1, S1>,
+      LongDecimal<P2, S2>>({name}, constraints);
+
+  // (short, short) -> short
+  registerFunction<
+      Func,
+      ShortDecimal<P3, S3>,
+      ShortDecimal<P1, S1>,
+      ShortDecimal<P2, S2>>({name}, constraints);
+
+  // (short, short) -> long
+  registerFunction<
+      Func,
+      LongDecimal<P3, S3>,
+      ShortDecimal<P1, S1>,
+      ShortDecimal<P2, S2>>({name}, constraints);
+
+  // (short, long) -> long
+  registerFunction<
+      Func,
+      LongDecimal<P3, S3>,
+      ShortDecimal<P1, S1>,
+      LongDecimal<P2, S2>>({name}, constraints);
+
+  // (long, short) -> long
+  registerFunction<
+      Func,
+      LongDecimal<P3, S3>,
+      LongDecimal<P1, S1>,
+      ShortDecimal<P2, S2>>({name}, constraints);
 }
 
-std::vector<std::shared_ptr<exec::FunctionSignature>>
-decimalMultiplySignature() {
-  return {exec::FunctionSignatureBuilder()
-              .integerVariable("a_precision")
-              .integerVariable("a_scale")
-              .integerVariable("b_precision")
-              .integerVariable("b_scale")
-              .integerVariable(
-                  "r_precision", "min(38, a_precision + b_precision + 1)")
-              .integerVariable(
-                  "r_scale",
-                  getResultScale(
-                      "a_precision + b_precision + 1", "a_scale + b_scale"))
-              .returnType("DECIMAL(r_precision, r_scale)")
-              .argumentType("DECIMAL(a_precision, a_scale)")
-              .argumentType("DECIMAL(b_precision, b_scale)")
-              .build()};
+std::string bounded(const std::string& value) {
```

Comment:
Is this used somewhere?

Reply:
I will delete it since it is used for allow precision loss false mode